Getting ready to Qt 6 (8/n). Use `QRegularExpression` in `AddressBookSortFilterProxyModel` class

hebasto opened the pull request:
This is a step in [migration](bitcoin/bitcoin/pull/24798) to Qt 6.

Related:
- bitcoin-core/gui#578
- bitcoin-core/gui#585

Based on https://github.com/bitcoin-core/gui/pull/591 and https://github.com/bitcoin-core/gui/pull/591, so only the last commit belongs to this PR.

promag commented on src/qt/addressbookpage.cpp:
```diff
@@ -46,12 +46,13 @@ class AddressBookSortFilterProxyModel final : public QSortFilterProxyModel
-
-        return true;
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 12, 0))
+        const auto pattern = filterRegularExpression().pattern();
```
Considering https://github.com/bitcoin-core/gui/pull/592/files#r878677419, I think the only required change is:
```cpp
#if (QT_VERSION >= QT_VERSION_CHECK(6, 0, 0))
    const auto pattern = filterRegularExpression()
#else
    const auto pattern = filterRegExp();
#endif
```

hebasto replied:
@promag 

Unfortunately, it does not work [for me](https://github.com/hebasto/gui/commits/220521-wc-demo) (

hebasto replied:
Thank you! [Updated.](https://github.com/bitcoin-core/gui/pull/593#issuecomment-1133658920)

promag reviewed (commented):
Concept ACK.

promag commented:
I think we can now close #592?

hebasto commented:
Updated c4e8af810c8ce07c98c0d5476c461a54c9f89c86 -> e280087946184b37c8a1eb345fae30ebb07f3384 ([pr593.01](https://github.com/hebasto/gui/commits/pr593.01) -> [pr593.02](https://github.com/hebasto/gui/commits/pr593.02)):

- implemented @promag's [suggestion](https://github.com/bitcoin-core/gui/pull/593#discussion_r878684298)

promag reviewed (approved):
Code review ACK e280087946184b37c8a1eb345fae30ebb07f3384.

Could also have a test case-insensitivity search.

hebasto commented:
> Could also have a test case-insensitivity search.

Could be a good first issue for following up :)

hebasto commented:
cc @luke-jr @prusnak @w0xlt

w0xlt reviewed (approved):
tACK https://github.com/bitcoin-core/gui/pull/593/commits/e280087946184b37c8a1eb345fae30ebb07f3384 on Ubuntu 21.10 Qt 5.15.2

Interesting note about `QRegularExpression` class:
https://doc.qt.io/qt-5/qregularexpression.html#notes-for-qregexp-users

hebasto commented:
@w0xlt 
> tACK [e280087](https://github.com/bitcoin-core/gui/commit/e280087946184b37c8a1eb345fae30ebb07f3384) on Ubuntu 21.10 Qt 5.15.2

Thank you for your review and testing!

FWIW, to test with Qt 6, the branch from bitcoin/bitcoin#24798 could be used.

promag reviewed (commented):
Tested ACK e280087946184b37c8a1eb345fae30ebb07f3384 with Qt6 on macOS 12 M1.

jarolrod reviewed (approved):
Tested ACK https://github.com/bitcoin-core/gui/commit/e280087946184b37c8a1eb345fae30ebb07f3384 on M1 mac, x86 mac, x86 Linux with Qt5 and separately with Qt6

The added tests in 5c5d8f2465be70cbc256ec59913ac0a3c7c958be are useful for verifying there's nothing being messed up by the changes. I also manually tested the address book for:
- The ability to still search effectively after the change
- That additions to the address book appear in the same order before and after the change
